system-addon/lib/ActivityStreamStorage.jsm

```diff
@@ -5,47 +5,70 @@ this.ActivityStreamStorage = class ActivityStreamStorage {
    * @param storeName String with the store name to access or array of strings
    *                  to create all the required stores
    */
-  constructor(storeName) {
+  constructor(storeNames, telemetry) {
     this.dbName = "ActivityStream";
     this.dbVersion = 3;
-    this.storeName = storeName;
+    this.storeNames = storeNames;
+    this.telemetry = telemetry;
   }
 
   get db() {
     return this._db || (this._db = this._openDatabase());
   }
 
-  async getStore() {
-    return (await this.db).objectStore(this.storeName, "readwrite");
+  async getStore(storeName) {
+    return (await this.db).objectStore(storeName, "readwrite");
   }
 
-  async get(key) {
-    return (await this.getStore()).get(key);
+  get(storeName, key) {
+    return this.requestWrapper(async () => (await this.getStore(storeName)).get(key));
   }
 
-  async getAll() {
-    return (await this.getStore()).getAll();
+  getAll(storeName) {
+    return this.requestWrapper(async () => (await this.getStore(storeName)).getAll());
   }
 
-  async set(key, value) {
-    return (await this.getStore()).put(value, key);
+  set(storeName, key, value) {
+    return this.requestWrapper(async () => (await this.getStore(storeName)).put(value, key));
   }
 
   _openDatabase() {
     return IndexedDB.open(this.dbName, {version: this.dbVersion}, db => {
       // If provided with array of objectStore names we need to create all the
       // individual stores
-      if (Array.isArray(this.storeName)) {
-        this.storeName.forEach(store => {
-          if (!db.objectStoreNames.contains(store)) {
-            db.createObjectStore(store);
-          }
-        });
-      } else if (!db.objectStoreNames.contains(this.storeName)) {
-        db.createObjectStore(this.storeName);
-      }
+      this.storeNames.forEach(store => {
+        if (!db.objectStoreNames.contains(store)) {
+          this.requestWrapper(() => db.createObjectStore(store));
+        }
+      });
     });
   }
+
+  getObjectStore(storeName) {
+    if (this.storeNames.includes(storeName)) {
+      return {
+        get: this.get.bind(this, storeName),
+        getAll: this.getAll.bind(this, storeName),
+        set: this.set.bind(this, storeName)
+      };
+    }
+
+    return null;
+  }
+
+  async requestWrapper(request) {
+    let result = null;
+    try {
+      result = await request();
+    } catch (e) {
+      if (this.telemetry) {
+        this.telemetry.handleUndesiredEvent({data: {event: "TRANSACTION_FAILED"}});
```

Comment:
Any way you could have this be slightly more specific within our privacy parameters?

Reply:
The database schema does not support additional information so anything we send will be dropped.